engines/macventure/gui.cpp

```diff
@@ -0,0 +1,1464 @@
+bool mainGameWindowCallback(Graphics::WindowClick, Common::Event &event, void *gui);
+bool outConsoleWindowCallback(Graphics::WindowClick, Common::Event &event, void *gui);
+bool selfWindowCallback(Graphics::WindowClick, Common::Event &event, void *gui);
+bool exitsWindowCallback(Graphics::WindowClick, Common::Event &event, void *gui);
+bool diplomaWindowCallback(Graphics::WindowClick, Common::Event &event, void *gui);
+bool inventoryWindowCallback(Graphics::WindowClick, Common::Event &event, void *gui);
+
+void menuCommandsCallback(int action, Common::String &text, void *data);
+
+Gui::Gui(MacVentureEngine *engine, Common::MacResManager *resman) {
+	_engine = engine;
+	_resourceManager = resman;
+	_windowData = NULL;
+	_controlData = NULL;
+	_draggedObj.id = 0;
+	_draggedObj.pos = Common::Point(0, 0);
+	_dialog = NULL;
+
+	_cursor = new Cursor(this);
+
+	_consoleText = new ConsoleText(this);
+	_graphics = NULL;
+
+	initGUI();
+}
+
+Gui::~Gui() {
+
+	if (_windowData)
+		delete _windowData;
+
+	if (_controlData)
+		delete _controlData;
+
+	if (_exitsData)
+		delete _exitsData;
+
+	if (_cursor)
+		delete _cursor;
+
+	if (_consoleText)
+		delete _consoleText;
+
+	if (_dialog)
+		delete _dialog;
+
+	clearAssets();
+
+	if (_graphics)
+		delete _graphics;
+}
+
+void Gui::initGUI() {
+	_screen.create(kScreenWidth, kScreenHeight, Graphics::PixelFormat::createFormatCLUT8());
+	_wm.setScreen(&_screen);
+
+	// Menu
+	_menu = _wm.addMenu();
+	if (!loadMenus())
+		error("GUI: Could not load menus");
+	_menu->setCommandsCallback(menuCommandsCallback, this);
+	_menu->calcDimensions();
+
+	loadGraphics();
+
+	if (!loadWindows())
+		error("GUI: Could not load windows");
+
+	initWindows();
+
+	assignObjReferences();
+
+	if (!loadControls())
+		error("GUI: Could not load controls");
+
+	draw();
+
+}
+
+void Gui::reloadInternals() {
+	clearAssets();
+	loadGraphics();
+}
+
+void Gui::draw() {
+	// Will be performance-improved after the milestone
+	_wm.setFullRefresh(true);
+
+	drawWindows();
+
+	_wm.draw();
+
+	drawDraggedObject();
+	drawDialog();
+	// TODO: When window titles with custom borders are in MacGui, this should be used.
+	//drawWindowTitle(kMainGameWindow, _mainGameWindow->getSurface());
+}
+
+void Gui::drawMenu() {
+	_menu->draw(&_screen);
+}
+
+void Gui::drawTitle() {
+	warning("drawTitle hasn't been tested yet");
+}
+
+void Gui::clearControls() {
+	if (!_controlData)
+		return;
+
+	Common::Array<CommandButton>::iterator it = _controlData->begin();
+	for (; it != _controlData->end(); ++it) {
+		it->unselect();
+	}
+}
+
+void Gui::initWindows() {
+	// Game Controls Window
+	_controlsWindow = _wm.addWindow(false, false, false);
+	_controlsWindow->setDimensions(getWindowData(kCommandsWindow).bounds);
+	_controlsWindow->setActive(false);
+	_controlsWindow->setCallback(commandsWindowCallback, this);
+	loadBorders(_controlsWindow, findWindowData(kCommandsWindow).type);
+
+	// Main Game Window
+	_mainGameWindow = _wm.addWindow(true, true, true);
```

Comment:
Should the main game window really be editable? (I'm getting a "beam" cursor in it, and this seems to be why.)